### DISCUSSION TITLE
All tensors are nan when loading the memory?

### DISCUSSION THREAD

**Discussion by nicholas0717** [Toni-SM/skrl#38] [Q&A] [answered]
Hi Toni!

Thank you for your great open-source library! Currently I'm trying to use skrl library to implement GAIL in conjunction with Isaacgym's Factory example tasks. But I met some troubles.

I loaded the best agent which was trained by PPO, and created a 100000 size memory to save the demonstration by going on the agent's eval. But I found that all the tensors are nan when I load the saved memory. Was there something wrong with my method?

Looking for your reply. Thanks! 

![image](https://user-images.githubusercontent.com/65715742/195978962-2afcb5e5-bcb4-4a5e-a41e-7e56e58b49c0.png)

**Comment by Toni-SM** [comment]
Hi @nicholas0717 

Thank you for giving the library a chance.

I have already identified the problem with memory.
Current trainer's implementations, in evaluation mode, call the `record_transition` agent's parent method (to write data to Tensorboard).
This implementation purposely avoids calling the agent method responsible for storing the data in memory during evaluation.

I started to fix it (I need to modify several files). 
I hope to solve it as soon as possible on Monday. I will reply here when I fix it.

**Reply by nicholas0717** [reply]
Thank you Toni!

I also found the reason that the `trainer.eval()` would call the `record_transition()` of agents/torch/base.py, not the `record_transition()` of agents/torch/ppo/ppo.py. This caused the memory I saved was empty. I'm looking forward to your good news.

By the way, I wonder that when will the new version with GAIL and HER release?

**Comment by Toni-SM** [comment, marked as answer]
Hi @nicholas0717 

I have made the changes to be able to store samples in memory during **evaluation**. You can find the changes in the `develop` branch (update the library from that branch)

Some considerations:
 - it is necessary to instantiate all involved models during the evaluation, not only the policy (if you want to use storage samples in the memory)
 - some memory tensors (such as *returns* and *advantages* for PPO, for example) will remain in `NaN` because they are only updated by the optimization algorithm (executed only during training)

Let me know if it works for you now :)

<hr>

Regarding the support for other agents/algorithms such as GAIL and HER, they are already on the roadmap for future skrl implementations (https://github.com/users/Toni-SM/projects/2/views/8). I hope to add them soon... need time... contributions are welcome :)

**Reply by nicholas0717** [reply]
Hi @Toni-SM 

Thank to your works. I have succeeded to store samples in the memory as the expert's demonstration.

Now I'm setting out to implement GAIL. Wish it smoothly.
